singularity-eos/eos/eos_sap_polynomial.hpp

```diff
@@ -0,0 +1,211 @@
+//------------------------------------------------------------------------------
+// © 2021-2023. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_EOS_SAP_POLYNOMIAL_HPP_
+#define _SINGULARITY_EOS_EOS_EOS_SAP_POLYNOMIAL_HPP_
+
+// stdlib
+#include <cmath>
+#include <cstdio>
+#include <string>
+
+// Ports-of-call
+#include <ports-of-call/portability.hpp>
+
+// Base stuff
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+#include <singularity-eos/base/robust_utils.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+namespace singularity {
+
+using namespace eos_base;
+
+class SAP_Polynomial : public EosBase<SAP_Polynomial> {
+ public:
+  SAP_Polynomial() = default;
+  PORTABLE_INLINE_FUNCTION
+  SAP_Polynomial(const Real rho0, const Real a0, const Real a1, const Real a2c,
+                 const Real a2e, const Real a3, const Real b0, const Real b1,
+                 const Real b2c, const Real b2e, const Real b3)
+      : _rho0(rho0), _a0(a0), _a1(a1), _a2c(a2c), _a2e(a2e), _a3(a3), _b0(b0), _b1(b1),
+        _b2c(b2c), _b2e(b2e), _b3(b3) {
+    checkParams();
+  }
+
+  SAP_Polynomial GetOnDevice() { return *this; }
+  PORTABLE_INLINE_FUNCTION void checkParams() const {
+    PORTABLE_ALWAYS_REQUIRE(_rho0 >= 0, "Reference density must be non-negative");
+  }
+  PORTABLE_INLINE_FUNCTION Real TemperatureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return 0.0;
```

Comment:
For the stubs, do we perhaps want to print a warning if compiled in debug mode?

Reply:
Is there an example of what you have in mind?

Reply:
Just
```C++
PORTABLE_WARN("This function is a stub.");
``
something like that. It should have been slurped in through the `portable_errors` code you're already using. Since it's missing the `ALWAYS` in the name, it will turn on only for debug builds.

Reply:
Perfect :)